Apps/Playground/Android/app/src/main/cpp/BabylonNativeJNI.cpp

```diff
@@ -8,182 +8,161 @@
-Java_BabylonNative_Wrapper_finishEngine(JNIEnv* env, jobject obj)
+extern "C"
 {
-    g_scriptLoader.reset();
-    g_inputBuffer.reset();
-    g_runtime.reset();
-    Babylon::Plugins::NativeEngine::DeinitializeGraphics();
-}
+    JNIEXPORT void JNICALL
+    Java_BabylonNative_Wrapper_initEngine(JNIEnv* env, jclass clazz)
+    {
+    }
 
-JNIEXPORT void JNICALL
-Java_BabylonNative_Wrapper_surfaceCreated(JNIEnv* env, jobject obj, jobject surface, jobject appContext)
-{
-    if (!g_runtime)
+    JNIEXPORT void JNICALL
+    Java_BabylonNative_Wrapper_finishEngine(JNIEnv* env, jclass clazz)
     {
-        g_runtime = std::make_unique<Babylon::AppRuntime>("");
+        g_scriptLoader.reset();
+        g_inputBuffer.reset();
+        g_runtime.reset();
+        Babylon::Plugins::NativeEngine::DeinitializeGraphics();
+    }
 
-        JavaVM* javaVM{};
-        if (env->GetJavaVM(&javaVM) != JNI_OK)
+    JNIEXPORT void JNICALL
+    Java_BabylonNative_Wrapper_surfaceCreated(JNIEnv* env, jclass clazz, jobject surface, jobject appContext)
+    {
+        if (!g_runtime)
         {
-            throw std::runtime_error("Failed to get Java VM");
-        }
+            g_runtime = std::make_unique<Babylon::AppRuntime>();
 
-        // TODO: This should be cleaned up via env->DeleteGlobalRef
-        auto globalAppContext = env->NewGlobalRef(appContext);
-
-        ANativeWindow* window = ANativeWindow_fromSurface(env, surface);
-        int32_t width  = ANativeWindow_getWidth(window);
-        int32_t height = ANativeWindow_getHeight(window);
-
-        g_runtime->Dispatch([javaVM, globalAppContext, window, width, height](Napi::Env env)
-        {
-            Babylon::Polyfills::Console::Initialize(env, [](const char* message, Babylon::Polyfills::Console::LogLevel level)
+            JavaVM* javaVM{};
+            if (env->GetJavaVM(&javaVM) != JNI_OK)
             {
-                switch (level)
-                {
-                case Babylon::Polyfills::Console::LogLevel::Log:
-                    __android_log_write(ANDROID_LOG_INFO, "BabylonNative", message);
-                    break;
-                case Babylon::Polyfills::Console::LogLevel::Warn:
-                    __android_log_write(ANDROID_LOG_WARN, "BabylonNative", message);
-                    break;
-                case Babylon::Polyfills::Console::LogLevel::Error:
-                    __android_log_write(ANDROID_LOG_ERROR, "BabylonNative", message);
-                    break;
-                }
-            });
-
-            Babylon::Platform::Initialize(env, javaVM, globalAppContext);
-
-            Babylon::Plugins::NativeWindow::Initialize(env, window, width, height);
+                throw std::runtime_error("Failed to get Java VM");
+            }
 
-            Babylon::Plugins::NativeEngine::InitializeGraphics(window, width, height);
-            Babylon::Plugins::NativeEngine::Initialize(env);
+            // TODO: This should be cleaned up via env->DeleteGlobalRef
```

Comment:
Could this be accomplished by giving the JavaScript a `Napi::External<gsl::final_action>` that calls the required function on JS shutdown?

Reply:
Perhaps. I didn't introduce this in this change. cc @ryantrem